pkg/asset/machines/baremetal/hosts.go

```diff
@@ -172,3 +194,104 @@ func Hosts(config *types.InstallConfig, machines []machineapi.Machine, userDataS
+	settings := &HostSettings{}
+
+	if config.Platform.BareMetal == nil {
+		return nil, fmt.Errorf("no baremetal platform in configuration")
+	}
+
+	// If Arbiter is not enabled, nothing happens
+	if !isArbiterEnabled(config) {
+		return nil, nil
+	}
+
+	numRequiredMasters := replicaCount(config.ControlPlane)
+	numMasters := 0
+
+	numRequiredArbiters := len(machines)
+	numArbiters := 0
+	for _, host := range config.Platform.BareMetal.Hosts {
+		// We make sure to keep an accurate count of known masters when explicitly set
+		if host.IsMaster() {
+			numMasters++
+		}
+
+		// If we know we're not an arbiter and the role isn't an empty value
+		// no further investigating is needed, skip creation.
+		if !host.IsArbiter() && host.Role != "" {
+			continue
+		}
+
+		// When it comes to hosts, the masters should have been generated first we skip those hosts
+		// and grab the hosts after to satisfy the number of required arbiters, and the host isn't
+		// explicitly set as an arbiter, this allows for mixed role types in the list.
+		if numMasters < numRequiredMasters && !host.IsArbiter() {
+			numMasters++
+			continue
+		}
+
+		if numArbiters < numRequiredArbiters {
+			secret, bmc := createSecret(host)
+			if secret != nil {
+				settings.Secrets = append(settings.Secrets, *secret)
+			}
+			newHost := createBaremetalHost(host, bmc)
+
+			if host.NetworkConfig != nil {
+				networkConfigSecret, err := createNetworkConfigSecret(host)
+				if err != nil {
+					return nil, err
+				}
+				settings.NetworkConfigSecrets = append(settings.NetworkConfigSecrets, *networkConfigSecret)
+				newHost.Spec.PreprovisioningNetworkDataName = networkConfigSecret.Name
+			}
+
+			// Setting CustomDeploy early ensures that the
+			// corresponding Ironic node gets correctly configured
+			// from the beginning.
+			newHost.Spec.CustomDeploy = &baremetalhost.CustomDeploy{
+				Method: "install_coreos",
+			}
+
+			newHost.ObjectMeta.Labels = map[string]string{
+				"installer.openshift.io/role": "control-plane",
+			}
+
+			// Link the new host to the currently available machine
+			machine := machines[numArbiters]
+			newHost.Spec.ConsumerRef = &corev1.ObjectReference{
+				APIVersion: machine.TypeMeta.APIVersion,
+				Kind:       machine.TypeMeta.Kind,
+				Namespace:  machine.ObjectMeta.Namespace,
+				Name:       machine.ObjectMeta.Name,
+			}
+			newHost.Spec.Online = true
+
+			// userDataSecret carries a reference to the arbiter ignition file
+			newHost.Spec.UserData = &corev1.SecretReference{Name: userDataSecret}
+			numArbiters++
+			settings.Hosts = append(settings.Hosts, newHost)
+		}
+	}
+
+	return settings, nil
+}
+
+// isArbiterEnabled Determine if arbiter is enabled based off of the install config.
+func isArbiterEnabled(config *types.InstallConfig) bool {
+	return config.Arbiter != nil &&
+		config.Arbiter.Replicas != nil &&
+		*config.Arbiter.Replicas > 0
+}
```

Comment:
nit/suggestion: might be more appropriate to define this as a pointer receiver on install config (or config.Arbiter if a pointer receiver can safely handle a nil receiver--not sure) especially if we reuse this logic

Reply:
Oh that's a great idea, I don't think I can add it to the Arbiter attribute since that's a MachinePool, but the install config makes a lot of sense. Updating